x/ecocredit/server/marketplace/buy_test.go

```diff
@@ -0,0 +1,240 @@
+				Quantity: "10", BidPrice: &ask, DisableAutoRetire: false, Expiration: &sellExp},
+		},
+	})
+	assert.NilError(t, err)
+
+	// order was filled, so sell order should no longer exist
+	_, err = s.marketStore.SellOrderTable().Get(s.ctx, sellOrderId)
+	assert.ErrorContains(t, err, ormerrors.NotFound.Error())
+	buyerBal, err := s.coreStore.BatchBalanceTable().Get(s.ctx, buyerAddr, 1)
+	assert.NilError(t, err)
+	retiredBal, err := math.NewDecFromString(buyerBal.Retired)
+	assert.NilError(t, err)
+	assert.Check(t, retiredBal.Equal(math.NewDecFromInt64(10)))
+}
+
+func TestBuy_Invalid(t *testing.T) {
+	t.Parallel()
+	s := setupBase(t)
+	_, _, buyerAddr := testdata.KeyTestPubAddr()
+	batchDenom := "C01-20200101-20200201-001"
+	start, end := timestamppb.Now(), timestamppb.Now()
+	ask := sdk.NewInt64Coin("ufoo", 10)
+	creditType := ecocredit.CreditType{
+		Name:         "carbon",
+		Abbreviation: "C",
+		Unit:         "tonnes",
+		Precision:    6,
+	}
+	testSellSetup(t, s, batchDenom, ask.Denom, ask.Denom[1:], "C01", start, end, creditType)
+	// make a sell order
+	any := gomock.Any()
+	s.paramsKeeper.EXPECT().GetParamSet(any, any).Do(func(any interface{}, p *ecocredit.Params) {
+		p.CreditTypes = []*ecocredit.CreditType{&creditType}
+	}).AnyTimes()
+	sellExp := time.Now()
+	res, err := s.k.Sell(s.ctx, &marketplace.MsgSell{
+		Owner: s.addr.String(),
+		Orders: []*marketplace.MsgSell_Order{
+			{BatchDenom: batchDenom, Quantity: "10", AskPrice: &ask, DisableAutoRetire: false, Expiration: &sellExp},
+		},
+	})
+	assert.NilError(t, err)
+	sellOrderId := res.SellOrderIds[0]
+
+	s.bankKeeper.EXPECT().HasBalance(any, any, any).Return(true).Times(3)
+
+	// sell order not found
+	_, err = s.k.Buy(s.ctx, &marketplace.MsgBuy{
+		Buyer: buyerAddr.String(),
+		Orders: []*marketplace.MsgBuy_Order{
+			{Selection: &marketplace.MsgBuy_Order_Selection{Sum: &marketplace.MsgBuy_Order_Selection_SellOrderId{SellOrderId: 532}},
+				Quantity: "10", BidPrice: &ask, DisableAutoRetire: false, Expiration: &sellExp},
+		},
+	})
+	assert.ErrorContains(t, err, ormerrors.NotFound.Error())
+
+	// exceeds decimal precision
+	_, err = s.k.Buy(s.ctx, &marketplace.MsgBuy{
+		Buyer: buyerAddr.String(),
+		Orders: []*marketplace.MsgBuy_Order{
+			{Selection: &marketplace.MsgBuy_Order_Selection{Sum: &marketplace.MsgBuy_Order_Selection_SellOrderId{SellOrderId: sellOrderId}},
+				Quantity: "10.23958230958", BidPrice: &ask, DisableAutoRetire: false, Expiration: &sellExp},
+		},
+	})
+	assert.ErrorContains(t, err, "exceeds maximum decimal places")
+
+	// mismatch auto retire settings
+	_, err = s.k.Buy(s.ctx, &marketplace.MsgBuy{
+		Buyer: buyerAddr.String(),
+		Orders: []*marketplace.MsgBuy_Order{
+			{Selection: &marketplace.MsgBuy_Order_Selection{Sum: &marketplace.MsgBuy_Order_Selection_SellOrderId{SellOrderId: sellOrderId}},
+				Quantity: "10", BidPrice: &ask, DisableAutoRetire: true, Expiration: &sellExp},
+		},
+	})
+	assert.ErrorContains(t, err, "auto-retire mismatch")
+
+}
```

Comment:
I'm not seeing tests for the following:

- expiration is in the future
- buyer has insufficient bank balance
- bid price ask price do not match
- bid price is less than ask price
- buy quantity is more than sell available
- partial fill disabled
- retirement location not provided (if auto-retire is true)

Reply:
> i think i'll actually move that expiration down into the Filter case, its only relevant there anyway

Makes sense.

> i think it might be better to change buy direct orders to ignore the DisablePartialFill field. the more i think about this though, theres quite a few fields not relevant to direct buy orders in the request... which seems like odd UX.. maybe we should just do a standalone BuyDirect RPC, rather than squashing it with filtered orders

I like the idea. I don't think we gave it enough thought on the product call. Maybe we can finish and merge this and do so in a separate pull request where we update the proto files. Wdyt? Can you create a followup issue?

> this would be covered by sdk.Msg ValidateBasic

We could check retirement location in `ValidateBasic` if auto-retire is disabled in the buy order but this would only be relevant if auto-retire is disabled in the sell order as well, which is why it was previously implemented in the server method.

For example, given a sell order with auto-retire enabled, when a user creates a buy order with auto-retire disabled and without a retirement location, then they would receive an error for empty retirement location. The user adds a retirement location and resubmits the message, then they would receive an error that they cannot disable auto-retire.

A think it would be a better user experience to check the retirement location after checking if the buyer is able to disabled auto-retire but we could go either way.

Reply:
> Can you create a followup issue? 

https://github.com/regen-network/regen-ledger/issues/905

as for the validate basic, i always lean towards failing as early as possible for simple message issues like this to avoid putting TX's in the mempool that we already know will fail